src/IC/Ref.hs

```
@@ -1142,6 +1158,33 @@ setAllTimesTo :: ICM m => Timestamp -> m ()
 setAllTimesTo ts = modify $
   \ic -> ic { canisters = M.map (\cs -> cs { time = ts }) (canisters ic) }
 
+runHeartbeat :: ICM m => CanisterId -> m ()
+runHeartbeat cid = do
+  can <- getCanister cid
+  if idleSinceLastHeartbeat (last_action can)
+  then return ()
+  else do
+    new_ctxt_id <- newCallContext $ CallContext
+      { canister = cid
+      , origin = FromHeartbeat
+      , responded = Responded False
```

I wonder if these call contexts should be considered “responded”. After all, we don’t expect a response from them.
```suggestion
      , responded = Responded True
```

---

This might also be relevant for the transition from stopping to stopped.

---

@nomeata Then the check in l. 634 is violated.
Heartbeat is invoked through `invokeEntry`, as a result a `ResponseMessage` is enqueued  in l. 640, then simply omitted in l. 725. It seems slightly indirect, but correct. What do you think?

---

Hmm, but why is there even a response message generated when the canister can't `reply`?

My theory: the call context looks starved (because of `Responded False` and no callbacks), and then the general starvation rule creates the response message. But on the phone right now, which means I may miss the full picture. 

In any case it isnt seemly for a reference implementation to do something such indirect, I'd say, if possible. Maybe there simply shouldn't be a call context when there is, after all, no call involved. Or we need different contexts. Not sure right now.

---

Thanks, I'll investigate further!

---

There should never be a `ResponseMessage` with a Hearbeat call context. I suggest to add  to `respondCallContext` a check
```
  when (origin ctxt == Heartbeat) $
    error "Internal error: Heartbeats cannot be responded to"
```
to fail early (similar to the double reply check there).

And I don’t see how `invokeEntry` would generate a response there. I think it was due to `starveCallContext` maybe.

---

Yes, `invokeEntry` has nothing to do with this. The thing is that we're attempting heartbeats also for stopped canisters.
In this case a `CallMessage` is rejected and a `ResponseMessage` is generated by `rejectCallContext`. At least that's my educated guess.

---

Maybe let's reject call contexts for heartbeats silently, i.e. make `rejectCallContext` conditional on the type of a call.

---

Hmm, stopped canisters, I see. Stopped or stopping?

I think the cleanest in this case is to simply not run heartbeats for stopped (and stopping) canisters, rather than oddly ignoring stuff later.

I hope some later refactoring will clean this up a bit. Some things don't feel right.

I wonder if the spec formalism has the same or similar bugs?

---

The spec is actually right - a canister needs to be running and not empty.

---

There are some additional preconditions for starving and deleting, too. I updated the code accordingly.